File: backend/controllers/recipeController.js
const Recipe = require('../models/Recipe');
const Rating = require('../models/Rating');

const createRecipe = async (req, res) => {
  try {
    console.log('Create recipe request:', req.body);
    console.log('File:', req.file);
    
    const { title, description, ingredients, steps, category, cookingTime } = req.body;
    
    const recipe = await Recipe.create({
      title,
      description,
      ingredients: JSON.parse(ingredients),
      steps: JSON.parse(steps),
      category,
      cookingTime,
      image: req.file ? req.file.filename : '',
      author: req.user._id
    });

    await recipe.populate('author', 'name');
    console.log('Recipe created:', recipe);
    res.status(201).json(recipe);
  } catch (error) {
    console.error('Create recipe error:', error);
    res.status(400).json({ message: error.message });
  }
};

const getRecipes = async (req, res) => {
  try {
    console.log('Search query params:', req.query);
    const { search, category, sort } = req.query;
    let query = {};

    if (search) {
      query.$or = [
        { title: { $regex: search, $options: 'i' } },
        { description: { $regex: search, $options: 'i' } },
        { ingredients: { $regex: search, $options: 'i' } }
      ];
    }

    if (category) {
      query.category = category;
    }

    let sortOption = { createdAt: -1 };
    if (sort === 'rating') {
      sortOption = { averageRating: -1 };
    }

    console.log('MongoDB query:', query);
    const recipes = await Recipe.find(query)
      .populate('author', 'name')
      .sort(sortOption);

    console.log('Found recipes:', recipes.length);
    res.json(recipes);
  } catch (error) {
    console.error('Search error:', error);
    res.status(400).json({ message: error.message });
  }
};

const getRecipe = async (req, res) => {
  try {
    const recipe = await Recipe.findById(req.params.id).populate('author', 'name');
    if (!recipe) {
      return res.status(404).json({ message: 'Recipe not found' });
    }
    res.json(recipe);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const updateRecipe = async (req, res) => {
  try {
    const recipe = await Recipe.findById(req.params.id);
    
    if (!recipe) {
      return res.status(404).json({ message: 'Recipe not found' });
    }

    if (recipe.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized' });
    }

    const { title, description, ingredients, steps, category, cookingTime } = req.body;
    
    recipe.title = title || recipe.title;
    recipe.description = description || recipe.description;
    recipe.ingredients = ingredients ? JSON.parse(ingredients) : recipe.ingredients;
    recipe.steps = steps ? JSON.parse(steps) : recipe.steps;
    recipe.category = category || recipe.category;
    recipe.cookingTime = cookingTime || recipe.cookingTime;
    
    if (req.file) {
      recipe.image = req.file.filename;
    }

    await recipe.save();
    await recipe.populate('author', 'name');
    
    res.json(recipe);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const deleteRecipe = async (req, res) => {
  try {
    const recipe = await Recipe.findById(req.params.id);
    
    if (!recipe) {
      return res.status(404).json({ message: 'Recipe not found' });
    }

    if (recipe.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized' });
    }

    await Recipe.findByIdAndDelete(req.params.id);
    res.json({ message: 'Recipe deleted' });
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

const getUserRecipes = async (req, res) => {
  try {
    const recipes = await Recipe.find({ author: req.user._id })
      .populate('author', 'name')
      .sort({ createdAt: -1 });
    
    res.json(recipes);
  } catch (error) {
    res.status(400).json({ message: error.message });
  }
};

module.exports = {
  createRecipe,
  getRecipes,
  getRecipe,
  updateRecipe,
  deleteRecipe,
  getUserRecipes
};

File: springboot-backend/src/main/java/com/foodstagram/controller/RecipeController.java
package com.foodstagram.controller;

import com.foodstagram.model.Recipe;
import com.foodstagram.model.User;
import com.foodstagram.repository.RecipeRepository;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

@RestController
@RequestMapping("/api/recipes")
public class RecipeController {
    private final RecipeRepository recipeRepository;
    private final ObjectMapper objectMapper;

    @Value("${upload.dir}")
    private String uploadDir;

    public RecipeController(RecipeRepository recipeRepository, ObjectMapper objectMapper) {
        this.recipeRepository = recipeRepository;
        this.objectMapper = objectMapper;
    }

    @GetMapping
    public ResponseEntity<?> getRecipes(
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) String sort) {
        List<Recipe> recipes;
        if ("rating".equals(sort)) {
            recipes = recipeRepository.searchRecipesByRating(search, category);
        } else {
            recipes = recipeRepository.searchRecipes(search, category);
        }
        return ResponseEntity.ok(recipes.stream().map(this::toMap).toList());
    }

    @GetMapping("/my-recipes")
    public ResponseEntity<?> getMyRecipes(@AuthenticationPrincipal User user) {
        List<Recipe> recipes = recipeRepository.findByAuthorOrderByCreatedAtDesc(user);
        return ResponseEntity.ok(recipes.stream().map(this::toMap).toList());
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getRecipe(@PathVariable Long id) {
        return recipeRepository.findById(id)
                .map(r -> ResponseEntity.ok(toMap(r)))
                .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<?> createRecipe(
            @RequestParam String title,
            @RequestParam String description,
            @RequestParam String ingredients,
            @RequestParam String steps,
            @RequestParam String category,
            @RequestParam Integer cookingTime,
            @RequestParam(required = false) MultipartFile image,
            @AuthenticationPrincipal User user) {
        try {
            Recipe recipe = new Recipe();
            recipe.setTitle(title);
            recipe.setDescription(description);
            recipe.setIngredients(objectMapper.readValue(ingredients, List.class));
            recipe.setSteps(objectMapper.readValue(steps, List.class));
            recipe.setCategory(category);
            recipe.setCookingTime(cookingTime);
            recipe.setAuthor(user);
            recipe.setAverageRating(0.0);
            recipe.setTotalRatings(0);

            if (image != null && !image.isEmpty()) {
                recipe.setImage(saveImage(image));
            }

            recipeRepository.save(recipe);
            return ResponseEntity.status(201).body(toMap(recipe));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updateRecipe(
            @PathVariable Long id,
            @RequestParam(required = false) String title,
            @RequestParam(required = false) String description,
            @RequestParam(required = false) String ingredients,
            @RequestParam(required = false) String steps,
            @RequestParam(required = false) String category,
            @RequestParam(required = false) Integer cookingTime,
            @RequestParam(required = false) MultipartFile image,
            @AuthenticationPrincipal User user) {
        try {
            Recipe recipe = recipeRepository.findById(id).orElse(null);
            if (recipe == null) return ResponseEntity.notFound().build();
            if (!recipe.getAuthor().getId().equals(user.getId()))
                return ResponseEntity.status(403).body(Map.of("message", "Not authorized"));

            if (title != null) recipe.setTitle(title);
            if (description != null) recipe.setDescription(description);
            if (ingredients != null) recipe.setIngredients(objectMapper.readValue(ingredients, List.class));
            if (steps != null) recipe.setSteps(objectMapper.readValue(steps, List.class));
            if (category != null) recipe.setCategory(category);
            if (cookingTime != null) recipe.setCookingTime(cookingTime);
            if (image != null && !image.isEmpty()) recipe.setImage(saveImage(image));

            recipeRepository.save(recipe);
            return ResponseEntity.ok(toMap(recipe));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(Map.of("message", e.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deleteRecipe(@PathVariable Long id, @AuthenticationPrincipal User user) {
        Recipe recipe = recipeRepository.findById(id).orElse(null);
        if (recipe == null) return ResponseEntity.notFound().build();
        if (!recipe.getAuthor().getId().equals(user.getId()))
            return ResponseEntity.status(403).body(Map.of("message", "Not authorized"));
        recipeRepository.delete(recipe);
        return ResponseEntity.ok(Map.of("message", "Recipe deleted"));
    }

    private String saveImage(MultipartFile file) throws Exception {
        File dir = new File(uploadDir);
        if (!dir.exists()) dir.mkdirs();
        String filename = System.currentTimeMillis() + "_" + file.getOriginalFilename();
        Path path = Paths.get(uploadDir + filename);
        Files.write(path, file.getBytes());
        return filename;
    }

    private Map<String, Object> toMap(Recipe r) {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("_id", r.getId());
        map.put("id", r.getId());
        map.put("title", r.getTitle());
        map.put("description", r.getDescription());
        map.put("ingredients", r.getIngredients());
        map.put("steps", r.getSteps());
        map.put("category", r.getCategory());
        map.put("cookingTime", r.getCookingTime());
        map.put("image", r.getImage());
        map.put("averageRating", r.getAverageRating());
        map.put("totalRatings", r.getTotalRatings());
        map.put("createdAt", r.getCreatedAt());
        map.put("author", Map.of("_id", r.getAuthor().getId(), "id", r.getAuthor().getId(), "name", r.getAuthor().getName()));
        return map;
    }
}
